apiserver/restrict_newer_client_test.go

```diff
@@ -1,164 +0,0 @@
-	c.Assert(caller, gc.IsNil)
-}
-
-func (r *restrictNewerClientSuite) TestReallyOldClientDisallowedMethod(c *gc.C) {
-	r.callerVersion.Major = jujuversion.Current.Major - 2
-	root := apiserver.TestingUpgradeOrMigrationOnlyRoot(true, r.callerVersion)
-	caller, err := root.FindMethod("Client", 3, "FullStatus")
-	c.Assert(errors.HasType[*params.IncompatibleClientError](err), jc.IsTrue)
-	c.Assert(caller, gc.IsNil)
-}
-
-func (r *restrictNewerClientSuite) TestReallyNewClientDisallowedMethod(c *gc.C) {
-	r.callerVersion.Major = jujuversion.Current.Major + 2
-	root := apiserver.TestingUpgradeOrMigrationOnlyRoot(true, r.callerVersion)
-	caller, err := root.FindMethod("Client", 3, "FullStatus")
-	c.Assert(errors.HasType[*params.IncompatibleClientError](err), jc.IsTrue)
-	c.Assert(caller, gc.IsNil)
-}
-
-func (r *restrictNewerClientSuite) TestAlwaysDisallowedMethod(c *gc.C) {
-	r.callerVersion.Major = jujuversion.Current.Major - 1
-	root := apiserver.TestingUpgradeOrMigrationOnlyRoot(true, r.callerVersion)
-	caller, err := root.FindMethod("ModelConfig", 3, "ModelSet")
-	c.Assert(errors.HasType[*params.IncompatibleClientError](err), jc.IsTrue)
-	c.Assert(caller, gc.IsNil)
-}
-
-func (r *restrictNewerClientSuite) TestWhitelistedClient(c *gc.C) {
-	// Ensure we're allowed to migrate from 2.9.x min release to 3.1.0.
-	r.assertWhitelistedClient(c, "2.9.41", "3.2.0", false)
-	r.assertWhitelistedClient(c, "2.9.42", "3.2.0", true)
-	r.assertWhitelistedClient(c, "2.9.42", "3.2.5", true)
-	r.assertWhitelistedClient(c, "2.9.42", "3.3.0", true)
-}
-
-func (r *restrictNewerClientSuite) assertWhitelistedClient(c *gc.C, callerVers, serverVers string, allowed bool) {
-	r.PatchValue(&jujuversion.Current, version.MustParse(serverVers))
-	r.callerVersion = version.MustParse(callerVers)
-	root := apiserver.TestingUpgradeOrMigrationOnlyRoot(true, r.callerVersion)
-	caller, err := root.FindMethod("ModelConfig", 3, "ModelSet")
-	if allowed {
-		c.Check(err, jc.ErrorIsNil)
-		c.Check(caller, gc.NotNil)
-	} else {
-		c.Check(err, gc.NotNil)
-		c.Check(caller, gc.IsNil)
-	}
-}
-
-func (r *restrictNewerClientSuite) TestAgentMethod(c *gc.C) {
-	// Ensure we're allowed to migrate from 2.9.x min release to 3.1.0.
-	r.assertAgentMethod(c, "2.9.42", "3.2.0", false)
-	r.assertAgentMethod(c, "2.9.43", "3.2.0", true)
-	r.assertAgentMethod(c, "2.9.43", "3.2.5", true)
-	r.assertAgentMethod(c, "2.9.43", "3.3.0", true)
-}
-
-func (r *restrictNewerClientSuite) assertAgentMethod(c *gc.C, agentVers, serverVers string, allowed bool) {
-	r.PatchValue(&jujuversion.Current, version.MustParse(serverVers))
-	r.callerVersion = version.MustParse(agentVers)
-	root := apiserver.TestingUpgradeOrMigrationOnlyRoot(false, r.callerVersion)
-	caller, err := root.FindMethod("Uniter", 18, "CurrentModel")
-	if allowed {
-		c.Check(err, jc.ErrorIsNil)
-		c.Check(caller, gc.NotNil)
-	} else {
-		c.Check(err, gc.NotNil)
-		c.Check(caller, gc.IsNil)
-	}
-}
```

Comment:
So if we wanted to be strictly as safe as possible, we *could* change this to know that "3.1.7" is the first version to actually check downgraded facades correctly.
However, IIRC, 3.1.x are actually compatible with 2.9 because they have facades that can be negotiated. So I *think* we're in "old clients don't check for low end facade versions, but it doesn't matter because they are compatible" and "new clients will check whether they can downgrade"